src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -103,6 +106,88 @@ internal sealed partial class LocalRewriter
+                    // As an optimization, ensure that assignments from temps to targets happen in the reverse order of effects:
+                    // temp1 = x;
+                    // temp2 = y;
+                    // temp3 = z;
+                    // c = temp3;
+                    // b = temp2;
+                    // a = temp1;
+
+                    // This makes it more likely that the stack optimizer pass will be able to eliminate the temps and replace them with stack push/pops.
+                    // Note also that no optimization can occur if the expression being converted is not a literal, e.g. `(a, b, c) = M()`. Order of access won't affect our ability to eliminate temps in that case.
+                    effects.assignments.ReverseContents();
+                }
+
+                visitedSymbols.Dispose();
+            }
+
+            static bool canReorderTargetAssignments(ArrayBuilder<Binder.DeconstructionVariable> targets, ref TemporaryArray<Symbol> visitedSymbols)
+            {
+                // If we know all targets refer to distinct variables, then we can reorder the assignments.
+                // We avoid doing this in any cases where aliasing could occur, e.g.:
+                // var y = 1;
+                // ref var x = ref y;
+                // (x, y) = (a, b);
+
+                foreach (var target in targets)
+                {
+                    if (target.Single is { } single)
+                    {
+                        if (single is not (BoundLocal { LocalSymbol.RefKind: RefKind.None }
+                                or BoundParameter { ParameterSymbol.RefKind: RefKind.None }
+                                or BoundDiscardExpression))
+                        {
+                            // This deconstruction assigns to a target which is not sufficiently simple.
+                            // We can't verify that the deconstruction does not use any aliases to variables.
+                            return false;
+                        }
+
+                        var expressionSymbol = single.ExpressionSymbol;
+                        Debug.Assert(expressionSymbol != null && expressionSymbol is not TypeSymbol);
+                        foreach (var visitedSymbol in visitedSymbols)
+                        {
+                            if ((object)visitedSymbol == expressionSymbol)
+                            {
+                                // This deconstruction writes to the same target multiple times, e.g:
+                                // (x, x) = (a, b);
+                                return false;
+                            }
+                        }
+                        visitedSymbols.Add(expressionSymbol);
+                    }
+
+                    if (target.NestedVariables is { } nestedVariables && !canReorderTargetAssignments(nestedVariables, ref visitedSymbols))
```

Comment:
>if (target.NestedVariables [](http://example.com/codeflow?start=20&length=26)

Is there a good reason in performing this check if ```if (target.Single is { } single)``` succeeded above?

Also, we checked ```target.Single``` and ```target.NestedVariables```. Is there an alternative state that ```target``` can have? If not, then consider structuring the code so that it is clear that these are the only states we can run into.

Reply:
Have corrected this to expect exactly one of Single or NestedVariables to be present.

Reply:
>> Is there an alternative state that target can have? If not, then consider structuring the code so that it is clear that these are the only states we can run into.  

> Have corrected this to expect exactly one of Single or NestedVariables to be present.

The code is written as though the ```target.NestedVariables is { } nestedVariables``` test can fail at this point. If that is not the case, then we can simply apply an unconditional cast in place of the first argument for the following ```canReorderTargetAssignments``` call. If the test can fail, then it looks like we not handling this situation and assuming that reordering can be done for it.